File: ts/test/proto-pollution.test.js
/* Copyright (c) 2013-2026 Richard Rodger and other contributors, MIT License */
'use strict'

const { describe, it, afterEach } = require('node:test')
const assert = require('node:assert')

const { util, Tabnas } = require('..')
const { deep } = util

// A JSON object gives `__proto__` (and `constructor`/`prototype`) as OWN
// enumerable keys — unlike a JS object literal, where `__proto__:` sets the
// prototype. These payloads are therefore built with JSON.parse to reproduce
// the attacker-supplied-JSON shape the merge functions must defend against.
const P = (s) => JSON.parse(s)

// The dangerous keys, in case a fix regresses.
const DANGER = ['__proto__', 'constructor', 'prototype']

describe('prototype-pollution', () => {

  // Belt-and-braces: if any assertion regresses and pollution leaks, remove
  // the sentinels so the rest of the suite is not poisoned.
  afterEach(() => {
    for (const sentinel of [
      'pwn', 'pwnOpts', 'pwnGram', 'pwnAlt', 'pwnNest', 'pwnDeep',
    ]) {
      delete Object.prototype[sentinel]
    }
  })

  it('options() with a JSON __proto__ payload does not pollute', () => {
    new Tabnas().options(P('{"__proto__":{"pwnOpts":1}}'))
    assert.strictEqual({}.pwnOpts, undefined)
    assert.strictEqual(
      Object.prototype.hasOwnProperty.call(Object.prototype, 'pwnOpts'),
      false,
    )
  })

  it('grammar({options:{__proto__:...}}) does not pollute', () => {
    new Tabnas().grammar(P('{"options":{"__proto__":{"pwnGram":1}}}'))
    assert.strictEqual({}.pwnGram, undefined)
  })

  it('grammar with __proto__ nested in alt data does not pollute', () => {
    const spec = P(
      '{"rule":{"top":{"open":[{"s":[1],"a":{"__proto__":{"pwnAlt":1}}}]}}}',
    )
    new Tabnas().grammar(spec)
    assert.strictEqual({}.pwnAlt, undefined)
  })

  it('all three dangerous keys, nested one level deep, are neutralized', () => {
    for (const k of DANGER) {
      const payload = P(`{"outer":{"${k}":{"pwnNest":1}}}`)
      new Tabnas().options(payload)
      assert.strictEqual(
        {}.pwnNest,
        undefined,
        `nested "${k}" polluted the prototype`,
      )
    }
  })

  it('deep() drops dangerous keys but keeps legitimate data', () => {
    // Direct unit-level proof on the merge primitive itself.
    for (const k of DANGER) {
      const base = {}
      deep(base, P(`{"${k}":{"pwnDeep":1}}`))
      assert.strictEqual({}.pwnDeep, undefined, `deep() merged "${k}"`)
      assert.strictEqual(
        Object.prototype.hasOwnProperty.call(base, k),
        false,
        `deep() wrote own key "${k}" onto base`,
      )
    }

    // A legitimate key that merely resembles a dangerous one must still merge.
    const merged = deep({ a: 1 }, { proto_thing: 2, constructor_name: 3 })
    assert.strictEqual(merged.a, 1)
    assert.strictEqual(merged.proto_thing, 2)
    assert.strictEqual(merged.constructor_name, 3)
  })

  it('a legitimate option named proto_thing still merges via options()', () => {
    const out = new Tabnas().options(P('{"proto_thing":7}'))
    assert.strictEqual(out.proto_thing, 7)
  })

})

File: go/proto_pollution_test.go
// Copyright (c) 2013-2026 Richard Rodger, MIT License

package tabnas

// Prototype-pollution guard. The TypeScript deep() merges attacker JSON and,
// without a guard, a `{"__proto__":{...}}` payload mutates Object.prototype
// globally. A Go map[string]any has no prototype, so a `__proto__` key
// decoded from JSON is only an ordinary string key here — but Deep is a port
// of deep(), so it drops `__proto__`/`constructor`/`prototype` identically,
// for parity and defense. These tests are the Go mirror of
// ts/test/proto-pollution.test.js.

import "testing"

var protoDangerKeys = []string{"__proto__", "constructor", "prototype"}

// Deep over plain maps must drop the dangerous keys (top level and nested
// one level deep) while keeping legitimate keys — including one whose name
// merely resembles a dangerous key.
func TestDeepMapDropsDangerousKeys(t *testing.T) {
	over := map[string]any{"proto_thing": 7, "constructor_name": 8}
	for _, k := range protoDangerKeys {
		over[k] = map[string]any{"pwn": 1}
	}
	got, ok := Deep(map[string]any{"keep": 1}, over).(map[string]any)
	if !ok {
		t.Fatalf("Deep returned %T, want map[string]any", Deep(map[string]any{}, over))
	}
	for _, k := range protoDangerKeys {
		if _, present := got[k]; present {
			t.Errorf("dangerous key %q survived the map merge", k)
		}
	}
	if got["keep"] != 1 || got["proto_thing"] != 7 || got["constructor_name"] != 8 {
		t.Errorf("legitimate keys were disturbed: %#v", got)
	}

	// Nested one level deep: the guard must fire at every recursion, not
	// only the top.
	for _, k := range protoDangerKeys {
		nested := Deep(
			map[string]any{},
			map[string]any{"outer": map[string]any{k: map[string]any{"pwn": 1}, "safe": 2}},
		).(map[string]any)
		inner, _ := nested["outer"].(map[string]any)
		if _, present := inner[k]; present {
			t.Errorf("nested dangerous key %q survived the map merge", k)
		}
		if inner["safe"] != 2 {
			t.Errorf("nested legitimate key dropped for %q case: %#v", k, inner)
		}
	}
}

// The OrderedMap merge path (the object node the parser actually builds) must
// drop the same keys and keep legitimate ones in order.
func TestDeepOrderedMapDropsDangerousKeys(t *testing.T) {
	over := NewOrderedMap()
	over.Set("proto_thing", 7)
	for _, k := range protoDangerKeys {
		over.Set(k, map[string]any{"pwn": 1})
	}
	got, ok := Deep(NewOrderedMap(), over).(*OrderedMap)
	if !ok {
		t.Fatalf("Deep returned %T, want *OrderedMap", Deep(NewOrderedMap(), over))
	}
	for _, k := range protoDangerKeys {
		if got.Has(k) {
			t.Errorf("dangerous key %q survived the ordered-map merge", k)
		}
	}
	if v, _ := got.Get("proto_thing"); v != 7 {
		t.Errorf("legitimate key proto_thing dropped: %v", v)
	}
}

// The JSON grammar/options door (GrammarSpecFromJSON + Grammar) must tolerate
// attacker payloads carrying the dangerous keys — in the options map, and
// nested in alt data — without error or panic, while a legitimately named
// rule (`proto_thing`) still installs and parses. Confirms the door routes
// through the guarded merge and rejects nothing legitimate.
func TestGrammarJSONDoorNeutralizesDangerousKeys(t *testing.T) {
	spec := `{"clear":true,` +
		`"options":{"rule":{"start":"top"},` +
		`"__proto__":{"pwn":1},"constructor":{"pwn":1},"prototype":{"pwn":1}},` +
		`"rule":{` +
		`"top":{"open":[{"s":"#NR","u":{"__proto__":{"pwn":1}}},{"p":"proto_thing"}]},` +
		`"proto_thing":{"open":[{"s":"#ST"}]}}}`

	j := loadJSONGrammar(t, spec)

	// The legitimately named rule installed and both branches parse.
	expectVerdicts(t, j, []string{"42", `"s"`}, nil)

	// A fresh merge after the door still carries no dangerous key: the guard
	// held, nothing global (a shared map) was mutated on the way through.
	fresh := Deep(map[string]any{}, map[string]any{"a": 1}).(map[string]any)
	for _, k := range protoDangerKeys {
		if _, present := fresh[k]; present {
			t.Errorf("post-door merge unexpectedly carries %q", k)
		}
	}
}
